bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/types/StateDescriptionFragmentBuilder.java

```diff
@@ -0,0 +1,239 @@
+            }
+
+            @Override
+            public @Nullable List<StateOption> getOptions() {
+                return options;
+            }
+
+            @Override
+            public @Nullable StateDescription toStateDescription() {
+                if (minimum == null && maximum == null && step == null && readOnly == null && pattern == null
+                        && options == null) {
+                    return null;
+                }
+                final Boolean ro = readOnly;
+                return new StateDescription(minimum, maximum, step, pattern, ro == null ? false : ro.booleanValue(),
+                        options);
+            }
+
+        };
+    }
+
+    /**
+     * Set the maximum for the resulting {@link StateDescriptionFragment}.
+     *
+     * @param maximum the maximum for the resulting {@link StateDescriptionFragment}.
+     * @return this builder.
+     */
+    public StateDescriptionFragmentBuilder withMaximum(BigDecimal maximum) {
+        this.maximum = maximum;
+        return this;
+    }
+
+    /**
+     * Set the minimum for the resulting {@link StateDescriptionFragment}.
+     *
+     * @param minimum the minimum for the resulting {@link StateDescriptionFragment}.
+     * @return this builder.
+     */
+    public StateDescriptionFragmentBuilder withMinimum(BigDecimal minimum) {
+        this.minimum = minimum;
+        return this;
+    }
+
+    /**
+     * Set the step for the resulting {@link StateDescriptionFragment}.
+     *
+     * @param step the step for the resulting {@link StateDescriptionFragment}.
+     * @return this builder.
+     */
+    public StateDescriptionFragmentBuilder withStep(BigDecimal step) {
+        this.step = step;
+        return this;
+    }
+
+    /**
+     * Set the pattern for the resulting {@link StateDescriptionFragment}.
+     *
+     * @param pattern the pattern for the resulting {@link StateDescriptionFragment}.
+     * @return this builder.
+     */
+    public StateDescriptionFragmentBuilder withPattern(String pattern) {
+        this.pattern = pattern;
+        return this;
+    }
+
+    /**
+     * Set readOnly for the resulting {@link StateDescriptionFragment}.
+     *
+     * @param readOnly readOnly for the resulting {@link StateDescriptionFragment}.
+     * @return this builder.
+     */
+    public StateDescriptionFragmentBuilder withReadOnly(Boolean readOnly) {
+        this.readOnly = readOnly;
+        return this;
+    }
+
+    /**
+     * Set the {@link StateOption}s for the resulting {@link StateDescriptionFragment}.
+     *
+     * @param options the {@link StateOption}s for the resulting {@link StateDescriptionFragment}.
+     * @return this builder.
+     */
+    public StateDescriptionFragmentBuilder withOptions(List<StateOption> options) {
+        this.options = options;
+        return this;
+    }
+
+    /**
+     * Merge the given {@link StateDescriptionFragment}. Set all unset ({@code null}) fields of this builder to the
```

Comment:
The fact that it only merges yet unfilled parameters is surprising - the intuitive way would be to use all non-null values of `fragment` and override `this`'s fields with them.

To me this is rather an indicator that this shouldn't be a `merge...()`-method but rather a `static create(SDF)`  which pre-initializes the builder with the given values.

Reply:
Thanks for pointing me to it. I had the feeling too but was not able to turn it into code ;-)

Reply:
For the `create(StateDescription legacy)` method your arguments hold true and I will refactor it that way.
For the StateDescriptionFragment merging I will try to refactor it in a way which is less surprising. Maybe an extra class or directly in the StateDescriptionService.